composite operator implementation.

jrrhuang opened the pull request:
Implement composite operator

jrrhuang committed:
composite operator implementation.

shiyunon commented on src/braket/circuits/circuit.py:
```diff
@@ -891,6 +892,24 @@ def copy(self) -> Circuit:
+        Call a decomposition pass across the circuit.
+
+        Returns:
+            Circuit: Circuit decomposed by one level.
```
Wonder if should support recursive decomposition.

shiyunon commented on src/braket/circuits/quantum_operator.py:
```diff
@@ -46,9 +46,6 @@ def __init__(self, qubit_count: int, ascii_symbols: Sequence[str]):
         if ascii_symbols is None:
             raise ValueError("ascii_symbols must not be None")
 
-        if len(ascii_symbols) != qubit_count:
```
It's safer to still check this for `Gate`, which means you might want to add this piece to the `__init__` method in Class `Gate`.

shiyunon commented on src/braket/circuits/composite_operators.py:
```diff
@@ -0,0 +1,345 @@
+    """
+
+    def __init__(self, precision_qubit_count: int, query_qubit_count: int, matrix: np.ndarray, condense=True):
+        verify_quantum_operator_matrix_dimensions(matrix)
```
I don't find the definition of this method.

jrrhuang replied:
This is in quantum_operator_helpers.

shiyunon replied:
How did you get it work without importing it?

jrrhuang replied:
I imported it when doing 'from braket.circuits.gates import *'

shiyunon commented on src/braket/circuits/composite_operator.py:
```diff
@@ -0,0 +1,50 @@
+    """
+
+    def __init__(self, qubit_count: int, ascii_symbols: Sequence[str]):
+        super().__init__(qubit_count=qubit_count, ascii_symbols=ascii_symbols)
```
It's not a problem of your repo. It looks like the `__init__` method in `super` which is `QuantumOperator` doesn't check if the input `ascii_symbols` is used by other quantum operators (so nothing you will have to do here).

shiyunon commented on src/braket/circuits/composite_operator.py:
```diff
@@ -0,0 +1,50 @@
+        """
+        raise NotImplementedError("to_ir has not been implemented yet.")
+
+    def decompose(self, target: QubitSet):
```
Not sure I like leaving this empty and define `decompose` in each `CompositeOperator` subclass (e.g. QFT, etc). `decompose` in each `CompositeOperator` subclass returns a list of instructions anyway, why not define a field `self.instructions` and let the superclass `CompositeOperator` return `self.instructions` then we don't have to overwrite  `decompose` methods in subclasses.

jrrhuang replied:
Hmm, so maybe I can do the decomposition in the circuit subroutine part (e.g. Circuit().qft) and store the instructions in self.instructions?

shiyunon replied:
I think you should have a `self.instructions` for every `CompositeOperator` then do decomposition here. In each subclasses, you just define the different `self.instructions`.

shiyunon replied:
Your implementation is not wrong. It's just that logically every `CompositeOperator` has a `self.instructions` and they should be decomposed the same way. What's different for each `CompositeOperator` is their specific `self.instructions`.

jrrhuang replied:
I'm having trouble figuring out a good way to do this since the only convenient place I can think of where I can add all the corresponding instructions for the quantum algorithm to `self.instructions` is the circuit subroutine method because we need to know the set of target qubits. But this doesn't work very well because some decompositions (e.g. QPE) use helper functions, which I can't access because the circuit subroutine method is a static method.

shiyunon replied:
Then it is the question of should it be "CompositeOperator" or "CompositeInstruction".

Also, can it be solved by attaching qubits to CompositeOperator and doing a qubit remapping when map operators to instructions? Haven't looked it into details.

dbolt replied:
I'm not sure I follow the original problem? The way it's implemented now the operator needs to be supplied the target qubits in order to produce the instructions. That means it either needs them in the constructor or via a function call (which is what is done today in this PR via the decompose function). If supplied in the constructor then the user has to supply the target to the Instruction and the Operator. For example: `circ = Circuit().add_instruction(GHZ([0,1,2]), [0,1,2]))`


If we have a `CompositeInstruction` then we have the issue of rendering the circuit since today printing is done by looping over the Instruction's operator ASCII symbols. We would need a single instance of Instruction to represent the aggregate instructions and would need to provide a way to print this composite instruction. This would require a bit of thinking to see how it could work. I'm not necessarily opposed to the idea, but would require a bit of work to see if it fits with the existing SDK design.

shiyunon commented on src/braket/circuits/circuit.py:
```diff
@@ -891,6 +892,24 @@ def copy(self) -> Circuit:
+        Call a decomposition pass across the circuit.
+
+        Returns:
+            Circuit: Circuit decomposed by one level.
```
1. I think the `Instruction` class should implement `decompose` and this method just loop over the instructions and call their `decompose`. Some of the logic you are doing here belongs to the `Instruction` class.
2. (Feature enhancement) You could support recursive decomposition as well, or at least giving users an option. For example, QFT can be recursively decompose: https://arxiv.org/pdf/quant-ph/0411069.pdf

jrrhuang replied:
Is the recursive decomposition you gave the same as the one as in here: https://github.com/aws/amazon-braket-examples/blob/main/examples/advanced_circuits_algorithms/QFT/QFT.ipynb.

Also, I think the option to recursively decompose should only be given when applicable right. For example, QFT should have this option, but not QPE.

shiyunon replied:
Yes, I think it is the same one.

The point is not recursive, but give users decompose levels. In your demo, you demoed:
```
circ.decompose().decompose()
```
A better UI might be `circ.decompose()` have a default level 1, which means your example becomes
```
circ.decompose(2)
```
And you will have something like
```
circ.decompose(level="infinite")
```
for recursive QFT.

Then for the choice of recursive QFT and your current QFT, you can either have
```
# Recursive QFT
circuit.qft_recursive()
# Your current QFT
circuit.qft()
```
Or
```
# Recursive QFT
circuit.qft(method="recursive")
# Your current QFT
circuit.qft()
```

In addition, users might want to just decompose certain Composite Operators, then it will be nice to have
```
circ.decompose(operators=["qft"], level="infinite")
```

shiyunon commented on src/braket/circuits/composite_operator.py:
```diff
@@ -0,0 +1,50 @@
+from braket.circuits.qubit_set import QubitSet
+
+
+class CompositeOperator(QuantumOperator):
```
Would love to hear from @dbolt. Here is one suggestion:

We have to ask the question we have been asking for the last 3 months again: what is the difference between a `CompositeOperator` and a `Gate`? Is it that `CompositeOperator` are decomposable while `Gate`s are not? I don't think so. `Gate`s should be decomposable as well: for example, we can decompose`cnot` to several`iswap` gates + 1q gates and vice versa. Actually, in Qiskit, all instructions are decomposable. Also, gate decomposition is actually important for nativization when translating a circuit to a native executable circuit on QPUs. Althought it's out of the scope of this PR to implement `decompose` for `Gate`, we still have to ask what is the difference.

 If we compare the subclasses of `Gate` and `CompositeOperator`:
```
class H(Gate):
    def __init__(self):
        super().__init__(qubit_count=1, ascii_symbols=["H"])
```

```
class QFT(CompositeOperator):
    def __init__(self, qubit_count):        
        super().__init__(qubit_count=qubit_count, ascii_symbols=["QFT"])
```
The difference is that `CompositeOperator` allows subclasses (e.g, QFT, QPE, etc) to construct operators with variable number of qubits, so it is actually a constructor or a template for a family of operators with different # of qubits.  On the other hand, `Gate` subclasses have fixed qubit count. Thus, the name `CompositeOperator` and the implemented class is not matching. I suggest the name to be our previous name `QuantumAlgorithm` (and we define `QuantumAlgorithm` to be a template for a family of circuits) or even something like `QuantumOperatorTemplate`. 

 Then, technically, the `QuantumAlgorithm` class should check all of its subclass obeys the rule: all subclasses should take variable # of qubits. This checking can be implemented in Python as follows:
```
import inspect
class QuantumAlgorithm(QuantumOperator):
    def __init_subclass__(cls):
        # This method get's called when a new subclass is created. 
        # The subclass name is passed as `cls`.
        if "qubit_count" not in inspect.signature(cls.__init__):
            raise Exception("A quantum algorithm should allow variable number of qubits.")
```

Actually, in the `Gate` class, we should have :
```
import inspect
class Gate(QuantumOperator):
    def __init_subclass__(cls):
        if "qubit_count" in inspect.signature(cls.__init__):
            raise Exception("A gate should have a fixed number of qubit(s).")
```
(The above change of `Gate` is probably beyond the this PR though).

dbolt replied:
tl;dr Pick a name for now since it'll have to be discussed again when merging into the main repository. The key is describing what it is and what it is not and then the name can follow suit.

I see this _thing_ as something that resolves into other `Operator` and has no dependency on the IR. Unlike Gates it doesn't produce IR, it only produces other `Operator`. Yes, Gates can be decomposed into other Gates in theory but they also produce IR since they're effectively python bindings to the IR layer. This _thing_ on the other hand does not since it doesn't have a dependency on the IR, it has a dependency on other `Operator`. I think of this _thing_ has 2 layers above the IR, CompositeOperator -> ... -> Gate -> IR.

jrrhuang committed:
Adjustments to gate and composite_operator. Add notebook to examples.

jrrhuang committed:
Notebook wording adjustment.

jrrhuang committed:
Wording change.

jrrhuang committed:
Correct GHZ composite operator.

jrrhuang committed:
Notebook correction.

jrrhuang committed:
Fix decompose.

dbolt commented on src/braket/circuits/ascii_circuit_diagram.py:
```diff
@@ -226,21 +227,42 @@ def _ascii_diagram_column(
-                # Set the margin to be a connector if not on the first qubit
-                if qubit != min(target_qubits):
-                    margins[qubit] = "|"
+            if hasattr(item, 'operator') and isinstance(item.operator, CompositeOperator):
+                spacing = max(1, len(item.operator.ascii_symbols[0])-2)
```
What assumptions are you making about the ascii symbols + CompositeOperator here? I don't see any validation in CompositeOperator that makes sure any assumptions are validated which is why I bring it up. What happens when someone implements a CompositeOperator that breaks these assumptions?

jrrhuang replied:
Is the problem the case when a composite operator doesn't have any ascii symbols?

dbolt commented on src/braket/circuits/gate.py:
```diff
@@ -39,6 +39,10 @@ def __init__(self, qubit_count: int, ascii_symbols: Sequence[str]):
             ValueError: `qubit_count` is less than 1, `ascii_symbols` are `None`, or
                 `ascii_symbols` length != `qubit_count`
         """
+        if len(ascii_symbols) != qubit_count:
+            msg = f"ascii_symbols, {ascii_symbols}, length must equal qubit_count, {qubit_count}"
+            raise ValueError(msg)
+
```
Missing test coverage for this. Before it was in QuantumOperator and now it was moved to gates, make sure a test reflects this.

dbolt commented on test/unit_tests/braket/circuits/test_composite_operators.py:
```diff
@@ -0,0 +1,312 @@
+import numpy as np
+import math
+import pytest
+from braket.ir.jaqcd.shared_models import (
+    Angle,
+    DoubleControl,
+    DoubleTarget,
+    MultiTarget,
+    MultiControl,
+    SingleControl,
+    SingleTarget,
+    TwoDimensionalMatrix,
+)
```
Why does the `test_composite_operators` module have a dependency on the IR but the implementing composite operator classes don't? Makes me think you're testing something that doesn't need to be tested.

I would expect your tests would assert that you are building the correct circuit / list of instructions. Why do you need to test you're building the right IR?

For example the GHZ has a dependency on the CNOT gate. If the CNOT gate changes what IR it produces it would break your test, even though none of your code changed. As long as you are building CNOT gates correctly your tests should be fine, it shouldn't matter how the CNOT is implemented and building the relevant IR.

jrrhuang replied:
Ok, that makes sense. I also saw that it is probably not good to restrain the types of input to only the IR subclasses (e.g. for Grover, it takes in a bitstring input, which is not an IR subclass). I've changed it so that it is no longer dependent on the IR.

jrrhuang committed:
Update for decomposition and recursive implementation:
- Add recursive qft, parameter for decomposition level.
- Validate len(ascii_symbols) is 1 for composite operator
- Removed IR dependency in unit test.

dbolt commented on src/braket/circuits/circuit.py:
```diff
@@ -892,21 +892,33 @@ def copy(self) -> Circuit:
         return self._copy()
 
-    def decompose(self) -> Circuit:
+    def decompose(self, level: Union[str, int]=1) -> Circuit:
```
- What happens if level is a negative number? Should you validate that get positive integers?
- What would it take to not accept one single string to imply fully recursive? Can you accept 0 to imply that instead?

dbolt commented on src/braket/circuits/circuit.py:
```diff
@@ -892,21 +892,33 @@ def copy(self) -> Circuit:
+
+        return circ
+
+    def contains_composite_operator(self):
```
- Does this need to be a public function? It doesn't look like anything needs this besides the decompose in this class. If it should be private then prefix it with an underscore. Why does this matter? It's python convention that anything prefixed with an underscore is private and can should not be invoked outside of the class.
- If this does need to be public then add a docstring and a return type for all public methods. Be sure to add that in to be consistent with the rest of the methods.

jrrhuang replied:
Yes, it is private, will fix.

dbolt commented on src/braket/circuits/composite_operator.py:
```diff
@@ -11,6 +11,19 @@ class CompositeOperator(QuantumOperator):
+                when printing a diagram of circuits.
+
+        Raises:
+            ValueError: `qubit_count` is less than 1, or `ascii_symbols` are `None`
```
Your code raises an error if `ascii_symbols` is not of length 1

dbolt commented on src/braket/circuits/composite_operator.py:
```diff
@@ -11,6 +11,19 @@ class CompositeOperator(QuantumOperator):
+        Raises:
+            ValueError: `qubit_count` is less than 1, or `ascii_symbols` are `None`
+        """
+        if len(ascii_symbols) != 1:
```
This is fine for now, but when we merge into the main repository it'll be worth thinking about how this can be changed to allow more ascii symbols.

dbolt commented on src/braket/circuits/composite_operators.py:
```diff
@@ -72,9 +72,17 @@ class QFT(CompositeOperator):
     """
 
-    def __init__(self, qubit_count):
+    def __init__(self, qubit_count, method="default"):
```
can you make `method` type based such that it's easier to validate and for clients to know what values are valid. Enums are a good approach to this. You can still allow a string version of the enum but it provides the value of clarity in terms of what input is allowed.

dbolt commented on src/braket/circuits/composite_operators.py:
```diff
@@ -72,9 +72,17 @@ class QFT(CompositeOperator):
 
     Args:
         qubit_count (int): Number of target qubits.
+        method (bool): Optional boolean flag for decomposition,
```
Method below is a string, not a boolean. Also update the docstring to reflect what valid values are.

dbolt commented on src/braket/circuits/composite_operators.py:
```diff
@@ -72,9 +72,17 @@ class QFT(CompositeOperator):
+    def __init__(self, qubit_count, method="default"):
+
+        if method != "default" and method != "recursive":
+            raise NameError("method must either be 'default' or 'recursive'.")
```
[NameError](https://docs.python.org/3.9/library/exceptions.html#NameError) doesn't really fit this use case. This feels more like a [TypeError](https://docs.python.org/3.9/library/exceptions.html#ValueError).

dbolt commented on test/unit_tests/braket/circuits/test_composite_operators.py:
```diff
@@ -1,29 +1,20 @@
-    (CompositeOperator.QFT, "qft", [MultiTarget], {}),
-    (CompositeOperator.iQFT, "iqft", [MultiTarget], {}),
-    (CompositeOperator.QPE, "qpe", [MultiTarget, MultiTarget, TwoDimensionalMatrix], {
+    (CompositeOperator.GHZ, "ghz", ["MultiTarget"], {}),
+    (CompositeOperator.QFT, "qft", ["MultiTarget"], {"option": "recursive"}),
+    (CompositeOperator.QFT, "qft", ["MultiTarget"], {"option": "default"}),
+    (CompositeOperator.mQFT, "mqft", ["MultiTarget"], {}),
+    (CompositeOperator.iQFT, "iqft", ["MultiTarget"], {}),
+    (CompositeOperator.QPE, "qpe", ["MultiTarget", "MultiTarget", "TwoDimensionalMatrix"], {
```
I think my feedback before about taking a dependency on the IR might have been misinterpreted. You still have a dependency on the IR because you are referencing things like "MultiTarget". You're testing that the GHZ decomposes into a set of IR which is the issue I'm trying to callout. What you should be testing is that GHZ decomposes into a set of Gates since that is what your code does. It doesn't matter how the gate turns into an IR though since that is up to the Gate implementation, not your classes implementation.

I'd update this test suite to confirm the behavior of GHZ, mQFT, QPE, etc. I can sync up with you offline if this feedback doesn't make any sense.

jrrhuang replied:
Yeah... sorry I'm not quite sure what you mean. I'm just using those labels (e.g. "MultiTarget) as a way to classify and retrieve the proper types of input that the composite operators should accept when running certain tests. Where are you referring to when you say that I'm testing that GHZ decomposed into a set of IR?
The only things that i set up tests for are: (1) whether the operator's circuit subroutine produces the same circuit as the circuit equivalent of its instruction, (2) if the operator's decomposition matches the expected circuit, and (3) if a ValueError is raised when we get mismatched target size.

jrrhuang committed:
- Clean up unit test for composite_operators.py
- Validate that 'level' parameter is non-negative for decompose(), and change infinite decomposition to level=0.

jrrhuang committed:
- Adjust test_composite_operators.py

jrrhuang committed:
Added more test cases to test_composite_operators.py

jrrhuang committed:
Simplify test_composite_operators.py

jrrhuang committed:
Slight simplification of two_dimensional_matrix_valid_input method implementation in test_composite_operators.py
